File: 07-steroid_langs/1-scala/src/main/scala/codely/contexts/rrss/post_likes/application/like/PostLiker.scala
package codely.contexts.rrss.post_likes.application.like

import codely.contexts.rrss.post_likes.domain.{PostLike, PostLikeRepository}
import codely.contexts.rrss.posts.application.find.PostFinder
import codely.contexts.rrss.users.application.find.UserFinder
import codely.contexts.rrss.posts.domain.PostDoesNotExistError
import codely.contexts.rrss.users.domain.UserDoesNotExistError
import codely.contexts.shared.domain.EventBus

class PostLiker(postFinder: PostFinder, userFinder: UserFinder, repository: PostLikeRepository, eventBus: EventBus) {
	def like(
		id: String,
		postId: String,
		likerUserId: String
	): Either[PostDoesNotExistError | UserDoesNotExistError, Unit] = {
		postFinder.find(postId).flatMap { _ =>
			userFinder.find(likerUserId).flatMap { _ =>
				val postLike = PostLike(id)

				repository.save(postLike)
				eventBus.publish(postLike.pullDomainEvents())

				Right(())
			}
		}
	}
}

File: 08-langs_particularities/1-java/src/main/java/tv/codely/contexts/rrss/post_likes/application/like/PostLiker.java
package tv.codely.contexts.rrss.post_likes.application.like;

import tv.codely.contexts.rrss.post_likes.domain.PostLike;
import tv.codely.contexts.rrss.post_likes.domain.PostLikeRepository;
import tv.codely.contexts.rrss.posts.application.find.PostFinder;
import tv.codely.contexts.rrss.posts.domain.PostDoesNotExistError;
import tv.codely.contexts.rrss.users.application.UserFinder;
import tv.codely.contexts.rrss.users.domain.UserDoesNotExistError;

public record PostLiker(PostFinder postFinder, UserFinder userFinder, PostLikeRepository repository) {
    public void like(String id, String postId, String likerUserId) throws PostDoesNotExistError, UserDoesNotExistError {
        postFinder.find(postId);
        userFinder.find(likerUserId);

        var postLike = PostLike.like(id, postId, likerUserId);

        repository.save(postLike);
    }
}
